engine/sshsock.go

```diff
@@ -11,42 +13,63 @@ import (
 	"google.golang.org/grpc/status"
 )
 
-const (
-	sshAuthSockEnv = "SSH_AUTH_SOCK"
-)
-
 // Map of id -> handler for that id.
-type MergedSocketProviders map[string]sshforward.SSHServer
+type SocketProvider struct {
+	Named NamedSocketProviders
+}
+
+type NamedSocketProviders map[string]sshforward.SSHServer
 
-func (m MergedSocketProviders) Register(server *grpc.Server) {
+func (m SocketProvider) Register(server *grpc.Server) {
 	sshforward.RegisterSSHServer(server, m)
 }
 
-func (m MergedSocketProviders) CheckAgent(ctx context.Context, req *sshforward.CheckAgentRequest) (*sshforward.CheckAgentResponse, error) {
+func (m SocketProvider) CheckAgent(ctx context.Context, req *sshforward.CheckAgentRequest) (*sshforward.CheckAgentResponse, error) {
 	id := sshforward.DefaultID
 	if req.ID != "" {
 		id = req.ID
 	}
-	h, ok := m[id]
+	if strings.HasPrefix(id, "socket:") {
+		// always allow SocketIDs
+		return &sshforward.CheckAgentResponse{}, nil
+	}
+	h, ok := m.Named[id]
 	if !ok {
 		return nil, status.Errorf(codes.NotFound, "no ssh handler for id %s", id)
 	}
 	return h.CheckAgent(ctx, req)
 }
 
-func (m MergedSocketProviders) ForwardAgent(stream sshforward.SSH_ForwardAgentServer) error {
+func (m SocketProvider) ForwardAgent(stream sshforward.SSH_ForwardAgentServer) error {
 	id := sshforward.DefaultID
 	opts, _ := metadata.FromIncomingContext(stream.Context()) // if no metadata continue with empty object
 	if v, ok := opts[sshforward.KeySSHID]; ok && len(v) > 0 && v[0] != "" {
 		id = v[0]
 	}
-	h, ok := m[id]
-	if !ok {
-		return status.Errorf(codes.NotFound, "no ssh handler for id %s", id)
+
+	var h sshforward.SSHServer
+	if key, socketID, ok := strings.Cut(id, ":"); key == "socket" && ok {
+		socket := core.NewSocket(core.SocketID(socketID))
+
+		var err error
+		h, err = socket.Server()
+		if err != nil {
+			return err
+		}
+	} else {
+		h, ok = m.Named[id]
+		if !ok {
+			return status.Errorf(codes.NotFound, "no ssh handler for id %s", id)
+		}
 	}
+
 	return h.ForwardAgent(stream)
 }
 
+const (
+	sshAuthSockEnv = "SSH_AUTH_SOCK"
```

Comment:
This isn't used in our new api at all right? If so, good to delete it now, users can accomplish the same thing by reading the SSH_AUTH_SOCK env var

Reply:
Was wondering about that; it's also used for `git { tree }`:

https://github.com/dagger/dagger/blob/bd17d70e8efca1d41206385546eebce74e9e4bbb/core/schema/git.go#L116

Plus a bunch of `project/` code.

https://github.com/dagger/dagger/blob/bd17d70e8efca1d41206385546eebce74e9e4bbb/project/project.go#L209

https://github.com/dagger/dagger/blob/bd17d70e8efca1d41206385546eebce74e9e4bbb/project/tsruntime.go#L37

https://github.com/dagger/dagger/blob/bd17d70e8efca1d41206385546eebce74e9e4bbb/project/pythonruntime.go#L62

For `git { tree }` I can just add a `tree(sshAuthSock: SocketID)` param, but it's less clear what to do for the project stuff. :thinking: Can it all just be ripped out?

Reply:
Yes feel free to rip the project parts of it out, not worth effort to maintain right now